services/brig/test/integration/API/User/Client.hs

```diff
@@ -498,6 +504,179 @@ testClientsWithoutPrekeys brig cannon db opts = do
+
+  post
+    ( brig
+        . paths ["users", "list-prekeys"]
+        . contentJson
+        . body (RequestBodyLBS $ encode userClients)
+        . zUser uid2
+    )
+    !!! do
+      const 200 === statusCode
+      const
+        ( Right $
+            expectedClientMap
+              domain
+              uid1
+              [ (clientId c11, Nothing),
+                (clientId c12, Just (unpackLastPrekey lk12))
+              ]
+              Nothing
+        )
+        === responseJsonEither
+  where
+    expectedClientMap :: Domain -> UserId -> [(ClientId, Maybe Prekey)] -> Maybe [Qualified UserId] -> QualifiedUserClientPrekeyMapV4
+    expectedClientMap domain u xs failed =
+      QualifiedUserClientPrekeyMapV4
+        { qualifiedUserClientPrekeys =
+            coerce $
+              mkQualifiedUserClientPrekeyMap $
+                Map.singleton domain $
+                  mkUserClientPrekeyMap $
+                    Map.singleton u $
+                      Map.fromList xs,
+          failedToList = failed
+        }
+
+testClientsWithoutPrekeysFailToListV4 :: Brig -> Cannon -> DB.ClientState -> Http ()
+testClientsWithoutPrekeysFailToListV4 brig cannon db = do
+  uid1 <- userId <$> randomUser brig
+  let (pk11, lk11) = (somePrekeys !! 0, someLastPrekeys !! 0)
+  c11 <- responseJsonError =<< addClient brig uid1 (defNewClient PermanentClientType [pk11] lk11)
+  let (pk12, lk12) = (somePrekeys !! 1, someLastPrekeys !! 1)
+  c12 <- responseJsonError =<< addClient brig uid1 (defNewClient PermanentClientType [pk12] lk12)
+
+  -- Simulating loss of all prekeys from c11 (due e.g. DB problems, business
+  -- logic prevents this from happening)
+  let removeClientKeys :: DB.PrepQuery DB.W (UserId, ClientId) ()
+      removeClientKeys = "DELETE FROM prekeys where user = ? and client = ?"
+  liftIO $
+    DB.runClient db $
+      DB.write removeClientKeys (DB.params DB.LocalQuorum (uid1, clientId c11))
+
+  uid2 <- fakeRemoteUser
+
+  let domain = Domain "foo.bar"
+
+  let userClients1 =
+        QualifiedUserClients $
+          Map.singleton domain $
+            Map.singleton uid1 $
+              Set.fromList [clientId c11, clientId c12]
+      userClients2 =
+        QualifiedUserClients $
+          Map.fromList
+            [ ( qDomain uid2,
+                Map.singleton (qUnqualified uid2) mempty
+              )
+            ]
+
+  WS.bracketR cannon uid1 $ \ws -> do
+    getClient brig uid1 (clientId c11) !!! do
+      const 200 === statusCode
+
+    post
+      ( brig
+          . paths ["users", "list-prekeys"]
+          . contentJson
+          . body (RequestBodyLBS $ encode userClients1)
+          . zUser (qUnqualified uid2)
+      )
+      !!! do
+        const 200 === statusCode
+        const
+          ( Right $
+              QualifiedUserClientPrekeyMapV4
+                { qualifiedUserClientPrekeys = QualifiedUserClientMap Map.empty,
+                  failedToList = pure [Qualified uid1 domain]
```

Comment:
As mentioned above, it seems wrong to me to qualify a local user `uid1` with a (fake) remote domain.